File: internal/interface/router/router.go
package router

import (
	"log/slog"
	"net/http"

	"github.com/go-chi/chi/v5"
	httpSwagger "github.com/swaggo/http-swagger"
	"github.com/swaggo/swag"

	"go-clean-template/internal/interface/handler"
	"go-clean-template/internal/interface/middleware"
)

func New(
	logger *slog.Logger,
	corsOrigin string,
	userHandler *handler.UserHandler,
	roleHandler *handler.RoleHandler,
) http.Handler {
	r := chi.NewRouter()

	r.Use(middleware.RequestID)
	r.Use(middleware.Recoverer(logger))
	r.Use(middleware.Logger(logger))
	r.Use(middleware.CORS(corsOrigin))

	r.Get("/health", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		w.Write([]byte(`{"status":"ok"}`))
	})

	r.Get("/openapi.json", func(w http.ResponseWriter, r *http.Request) {
		jsonStr, err := swag.ReadDoc()
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		w.Write([]byte(jsonStr))
	})

	r.Get("/swagger", func(w http.ResponseWriter, r *http.Request) {
		http.Redirect(w, r, "/swagger/", http.StatusMovedPermanently)
	})
	r.Get("/swagger/*", httpSwagger.Handler(
		httpSwagger.URL("/swagger/doc.json"),
	))

	r.Route("/api/v1", func(r chi.Router) {
		r.Route("/users", func(r chi.Router) {
			r.Get("/", userHandler.GetAll)
			r.Post("/", userHandler.Create)
			r.Get("/{id}", userHandler.GetByID)
			r.Put("/{id}", userHandler.Update)
			r.Delete("/{id}", userHandler.Delete)
			r.Get("/{id}/roles", userHandler.GetUserRoles)
			r.Post("/{id}/roles", userHandler.AssignRoles)
		})

		r.Route("/roles", func(r chi.Router) {
			r.Get("/", roleHandler.GetAll)
			r.Post("/", roleHandler.Create)
			r.Get("/{id}", roleHandler.GetByID)
			r.Put("/{id}", roleHandler.Update)
			r.Delete("/{id}", roleHandler.Delete)
		})
	})

	return r
}

File: src/interface/router.rs
use std::sync::Arc;

use axum::routing::{delete, get, post, put};
use axum::{Json, Router};

use crate::interface::handler::role::RoleHandler;
use crate::interface::handler::user::UserHandler;
use crate::interface::middleware;

pub fn new(user_handler: UserHandler, role_handler: RoleHandler) -> Router {
    let user_handler = Arc::new(user_handler);
    let role_handler = Arc::new(role_handler);

    let user_routes = Router::new()
        .route("/", get(UserHandler::list_all))
        .route("/", post(UserHandler::create))
        .route("/{id}", get(UserHandler::get_by_id))
        .route("/{id}", put(UserHandler::update))
        .route("/{id}", delete(UserHandler::delete))
        .route("/{id}/roles", get(UserHandler::get_roles))
        .route("/{id}/roles", post(UserHandler::assign_roles))
        .with_state(user_handler);

    let role_routes = Router::new()
        .route("/", get(RoleHandler::list_all))
        .route("/", post(RoleHandler::create))
        .route("/{id}", get(RoleHandler::get_by_id))
        .route("/{id}", put(RoleHandler::update))
        .route("/{id}", delete(RoleHandler::delete))
        .with_state(role_handler);

    Router::new()
        .route("/health", get(|| async {
            Json(serde_json::json!({"status": "ok"}))
        }))
        .nest("/api/v1/users", user_routes)
        .nest("/api/v1/roles", role_routes)
        .layer(axum::middleware::from_fn(middleware::request_id))
        .layer(axum::middleware::from_fn(middleware::recover_middleware))
        .layer(axum::middleware::from_fn(middleware::logger_middleware))
        .layer(middleware::cors_layer())
}
